File: spiral-matrix/spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        
        vector<int> result;
        if(matrix.empty()) return result;
        
        int m=matrix.size();
        int n=matrix[0].size();
        
        int top=0;
        int right=n-1;
        int bottom=m-1;
        int left=0;
        
        while(top<=bottom && left<= right){
            //top
            for(int i=left;i<=right;i++){
                result.push_back(matrix[top][i]);
            }
            top++;
            
            if(top<=bottom && left <= right){
                //right
                for(int i=top;i<=bottom;i++){
                    result.push_back(matrix[i][right]);
                }
                
            }
            
            right--;
            
            //bottom
            if(top<=bottom && left <= right){
                //right
                for(int i=right;i>=left;i--){
                    result.push_back(matrix[bottom][i]);
                }
                
            }
             
            bottom--;
            
            if(top<=bottom && left <= right){
                //left
                for(int i=bottom; i>=top; i--){
                    result.push_back(matrix[i][left]);
                }
            }
        
            left++;
        }
        return result;
    }
};

File: 54-spiral-matrix/spiral-matrix.py
class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """

        m = len(matrix)
        n = len(matrix[0]) 

        left = 0
        right = n - 1
        top = 0
        bottom = m - 1

        result = []
        idx = 0

        while((top <= bottom) and (left<= right)):
            # top row 
            for i in range(left, right + 1, 1):
                result.append(matrix[top][i])

            top += 1
            
            if (top<=bottom and left<= right):
                # right column
                for i in range(top, bottom + 1, 1):
                    result.append(matrix[i][right])

            right -= 1

            if (top<=bottom and left<= right):
                # bottom row
                for i in range(right, left-1, -1):
                    result.append(matrix[bottom][i])

            bottom -= 1

            if (top<=bottom and left<= right):
                # left column
                for i in range(bottom, top-1, -1):
                    result.append(matrix[i][left])
            
            left += 1

        return result

